datalad/distribution/tests/test_update.py

```diff
@@ -442,3 +446,165 @@ def test_multiway_merge(path):
+    ds_src_s1 = ds_src_s0.create("s1")
+    ds_src.save(recursive=True)
+    if on_adjusted:
+        # Note: We adjust after creating all the datasets above to avoid a bug
+        # fixed in git-annex 7.20191024, specifically bbdeb1a1a (sync: Fix
+        # crash when there are submodules and an adjusted branch is checked
+        # out, 2019-10-23).
+        for ds in [ds_src, ds_src_s0, ds_src_s1]:
+            _adjust(ds.repo)
+        ds_src.save(recursive=True)
+    assert_repo_status(ds_src.path)
+
+    ds_clone = install(source=ds_src.path, path=path / "clone",
+                       recursive=True, result_xfm="datasets")
+    ds_clone_s1 = Dataset(ds_clone.pathobj / "s0" / "s1")
+
+    ds_src_s1.repo.checkout("master^0")
+    (ds_src_s1.pathobj / "foo").write_text("foo content")
+    ds_src.save(recursive=True)
+    assert_repo_status(ds_src.path)
+
+    ds_clone.update(merge=True, recursive=True)
+    assert_repo_status(ds_clone.path)
+    if not on_adjusted:
+        # We brought in the revision and got to the same state of the remote.
+        # Blind saving here without bringing in the current subdataset revision
+        # would have resulted in a new commit in ds_clone that reverting the
+        # last subdataset ID recorded in ds_src.
+        eq_(ds_clone.repo.get_hexsha(), ds_src.repo.get_hexsha())
+
+    # Record a revision in the parent and then move HEAD away from it so that
+    # the explicit revision fetch fails.
+    (ds_src_s1.pathobj / "bar").write_text("bar content")
+    ds_src.save(recursive=True)
+    ds_src_s1.repo.checkout(
+        ds_src_s1.repo.get_corresponding_branch("master"))
+    # This is the default, but just in case:
+    ds_src_s1.repo.config.set("uploadpack.allowAnySHA1InWant", "false",
+                              where="local")
+    res = ds_clone.update(merge=True, recursive=True, on_failure="ignore")
+    if on_adjusted:
+        assert_not_in_results(res, status="error")
+    else:
+        # The fetch with the explicit ref fails because it isn't advertised.
+        assert_in_results(
+            res,
+            status="impossible",
+            path=ds_clone_s1.path,
+            action="update")
+
+    # Back to the detached head.
+    ds_src_s1.repo.checkout("HEAD@{1}")
+    # Set up a case where update() will not resolve the sibling.
+    ds_clone_s1.repo.call_git(["branch", "--unset-upstream"])
+    ds_clone_s1.config.reload(force=True)
+    ds_clone_s1.repo.call_git(["remote", "add", "other", ds_src_s1.path])
+    if on_adjusted:
+        assert_not_in_results(res, status="error")
+    else:
+        # In this case, update() won't abort if we call with merge=False, but
+        # it does if the revision wasn't brought down in the `fetch(all_=True)`
+        # call.
+        assert_in_results(
+            ds_clone.update(recursive=True, on_failure="ignore"),
+            status="impossible",
+            path=ds_clone_s1.path,
+            action="update")
```

Comment:
I ran these tests locally on top of gh-4087 (6e6210082), and this assertion fails.  A commit here adds a `fetch()` call that catches a `GitCommandError`, but that will be a `CommandError` following gh-4087.  So, assuming that gh-4087 will be merged first, this PR will need to be adjusted.